### DISCUSSION TITLE
Using C# dev kit in a devcontainer

### DISCUSSION THREAD

**Discussion by ingentismg** [devcontainers/community#111] [Q&A]
Hello,

I am currently trying to setup a devcontainer for use with an existing .dotnet 8 project and the VSCode _C# dev kit_ extension. Sadly I am constantly running into problems.

My current setup is based on https://github.com/devcontainers/images/blob/main/src/dotnet/.devcontainer/:
Dockerfile
``` Dockerfile
# See https://github.com/devcontainers/images/blob/main/src/dotnet/.devcontainer/Dockerfile
ARG VARIANT=8.0.101-alpine3.19
FROM mcr.microsoft.com/dotnet/sdk:${VARIANT}
ENV PATH $PATH:/home/vscode/.dotnet:/home/vscode/.dotnet/tools

# clear this environment variable so xml docs from NuGet packages are unpackaged. The default dotnet/sdk image sets it to 'skip'.
# see https://github.com/dotnet/dotnet-docker/issues/2790
ENV NUGET_XMLDOC_MODE=
```

docker-compose.yaml excerpt (I know some things are not ideal but I have reworked a lot, some commands could now be included in the Dockerfile)
```yaml
services:
  dotnet8-app:
    build:
      context: ./dotnet8-container
      args:
        - VARIANT=8.0.101-alpine3.19
    stop_grace_period: 1m
    ports:
      - "3001:3001"
    expose:
      - 3001
    volumes:
      - ../.:/workspace:cached
    command: [sh, -c, "apk add --no-cache bash openssh && git config --global --add safe.directory /workspace && trap : TERM INT; sleep infinity & wait"]
```

devcontainer.json
```json
{
    "name": "dotnet8 container",
    "dockerComposeFile": [
        "../docker-compose.yaml",
        "../docker-compose.development.yaml" // that is just mounting the dotnet userSecrets
    ],
    "service": "dotnet8-app",
    "runServices": [
        "dotnet8-app",
        "sql"
    ],
    "remoteUser": "vscode",
    "workspaceFolder": "/workspace/apps",
    "customizations": {
        // Configure properties specific to VS Code.
        "vscode": {
            // Add the IDs of extensions you want installed when the container is created.
            "extensions": [
                "ms-dotnettools.csdevkit"
            ],
            "settings": {
                "telemetry.enableTelemetry": false,
                "dotnetAcquisitionExtension.enableTelemetry": false
            }
        }
    },
    "features": {
        "ghcr.io/devcontainers/features/common-utils:2": {
            "username": "vscode",
            "userUid": "1000",
            "userGid": "1000"
        }
    }
}
```

The problems that I have:
1. The example shown in https://github.com/devcontainers/images/blob/main/src/dotnet/.devcontainer/Dockerfile is adding `/home/vscode/.dotnet/tools` to the `PATH` and I do not understand why. That folder does not exist.
I know that I am installing a different extension, `ms-dotnettools.csdevkit` instead of `ms-dotnettools.csharp`, but that should be installed automatically by the dev kit anyways. Even if I try `ms-dotnettools.csharp` directly, that _tools_ folder does not exist.
Is that an outdated line or is something missing that should create this folder?

1. I want to use the `"remoteUser": "vscode"` option, as that will solve a lot of file permission problems. The problem is that the C# dev kit tries to run `dotnet workload update` but can't: `Inadequate permissions. Run the command with elevated privileges.`
How can I run VSCode as the user and still have dotnet be able to update the workload? Is this connected to the missing _tools_ in the user home?

1. Roughly half the time I start the container the C# dev kit extension will crash and I have to restart the container until it just works again. Is there something wrong with the devcontainer or the extension (in which case I will move this to the extension repository)?
```
Starting Spawn .NET server...
Starting Open a solution...
Starting Open a solution with environment service...
Starting Clear environment...
Using local .NET runtime at "/home/vscode/.vscode-server/data/User/globalStorage/ms-dotnettools.vscode-dotnet-runtime/.dotnet/7.0.15~x64/dotnet"
Using runtime installed in SDK.
.NET server STDOUT: [createdump] Gathering state for process 467 Microsoft.Visua
[createdump] Crashing thread 01e3 signal 11 (000b)
...
```

Could you maybe add an additional dotnet example showing the usage of the C# dev kit with devcontainers?
Any help is appreciated. If some questions/problems are better suited as an issue in the vscode-dotnettools repo I will move there.

Thank you and best regards,
Marcus

**Comment by ingentismg** [comment]
If that is important:
The project has been cloned into WSL2 (Ubuntu).

**Comment by samruddhikhandale** [comment]
Hello Marcus 👋 

1. The `/home/vscode/.dotnet/tools` directory is typically used for .NET Global Tools. These are special NuGet packages that contain a console application. The `dotnet tool install -g` command installs a tool globally in this directory. If you're not using any .NET Global Tools, this directory might not exist. It's added to the PATH so that any installed tools can be run from any location.
2. The `dotnet workload update` command requires elevated privileges because it modifies system-wide locations. Running as a non-root user, you won't have the necessary permissions. One solution is to use a Dockerfile to run these commands as root during the image build process. (I haven't tested this, so let me know if this helps ).

Here's an example:

```
# Switch to root
USER root

# Run the command with elevated privileges
RUN dotnet workload update

# Switch back to the vscode user
USER vscode
```

3. The crash you're experiencing could be due to a variety of factors. It might be a problem with the extension, the .NET SDK, or the dev container setup. It's hard to say without more information. I recommend opening an issue in the vscode-dotnettools repo, providing all the details you've shared here.

**Reply by ingentismg** [reply]
Hi Samruddhi,

thanks for the response.

1. Ohh, I see. I am new to .dotnet so I did not know that.
2. I made other attempts to try something similar as well. Sadly, from my limited tests, the extension will still try to run the command in the user context as well.
3. In an Ubuntu container it is crashing with memory errors. So I guess this is more related to the extension than the devcontainers.

This very much sounds like the extension has to fix these things to be compatible with devcontainers.
Sadly due to other missing features I already gave up on it (e.g. even as root it won't install all dependencies properly).

I appreciate the help anyways.
Best regards,
Marcus
